python/iceberg/parquet/parquet_reader.py

```diff
@@ -0,0 +1,240 @@
+
+        self._case_sensitive = case_sensitive
+        if start is not None or end is not None:
+            raise NotImplementedError("Partial file reads are not yet supported")
+            # self.start = start
+            # self.end = end
+
+        self.materialized_table = False
+        self.curr_iterator = None
+        self._table = None
+        self._df = None
+
+        _logger.debug("Reader initialized for %s" % self._input.path)
+
+    @property
+    def stats(self) -> dict:
+        return dict(self._stats)
+
+    def to_pandas(self) -> Union[pd.Series, pd.DataFrame]:
+        if not self.materialized_table:
+            self._read_data()
+            with profile("to_pandas", self._stats):
+                if self._table is not None:
+                    self._df = self._table.to_pandas(use_threads=True)
+                else:
+                    self._df = None
+
+        return self._df
+
+    def to_arrow_table(self) -> pa.Table:
+        if not self.materialized_table:
+            self._read_data()
+
+        return self._table
+
+    def _read_data(self) -> None:
+        _logger.debug("Starting data read")
+
+        # only scan the columns projected and in our file
+        cols_to_read = prune_columns(self._file_schema, self._expected_schema)
+
+        with profile("read data", self._stats):
+            arrow_dataset = ds.FileSystemDataset.from_paths([self._input.location()],
+                                                            schema=self._arrow_file.schema_arrow,
+                                                            format=ds.ParquetFileFormat(),
+                                                            filesystem=fs.LocalFileSystem())
+
+            arrow_table = arrow_dataset.to_table(columns=cols_to_read, filter=self._filter)
+
+        # process schema evolution if needed
+        with profile("schema_evol_proc", self._stats):
+            processed_tbl = self.migrate_schema(arrow_table)
+            for i, field in self.get_missing_fields():
+                dtype_func = DTYPE_MAP.get(field.type.type_id)
+                if dtype_func is None:
+                    raise RuntimeError("Unable to create null column for type %s" % field.type.type_id)
+
+                dtype = dtype_func(field)
+                processed_tbl = (processed_tbl.add_column(i,
+                                                          pa.field(field.name, dtype[0], True, None),
+                                                          ParquetReader.create_null_column(processed_tbl[0],
+                                                                                           dtype)))
+        self._table = processed_tbl
+        self.materialized_table = True
+
+    def __iter__(self):
+        return self
+
+    def __next__(self):
+        if self.curr_iterator is None:
+            if not self.materialized_table:
```

Comment:
Is there any way to do this w/o first going to pandas? Pandas is the most expensive part of this operation and if we could avoid it (say get tuples direct from pyarrow) performance would significantly improve

Reply:
Let me revisit this.  Internally(at NFLX), we generally end up in a pandas DF, so that was the motivation here, but as arrow has matured, it does seem like it's more viable to leave things in an arrow buffer, and let clients deal with how they want to post-process.

Reply:
@rymurr I updated this, although I'm wondering if it makes more sense to remove this iterator logic and to_pandas from the parquet reader and move it to a higher level class because above the parquet reader, we will need something that combines all the various file scans into a single arrow table/pandas df.

Reply:
I agree. I think keeping this at the arrow level is best and having a layer above which concats the arrow tables together before handing back to the client. 

This of course could lead to memory difficulties so a future enhancement might be to do some sort of chunked reads (moving to arrow's datasets is preferable to writing our own tho :-)